### ISSUE TITLE
Add fast-memory-mapping alternative(s) to vaex

### ISSUE THREAD

**Issue by minouHub** [radis/radis#658]
### Description
Vaex is a fast memory mapping engine; better than Pandas for Dataframe management. However it is not available in `Python >=3.11` as of June 2024, (https://github.com/vaexio/vaex/issues/2397). **This issue/to-do proposes to implement an alternative to `vaex`.** 

The user would be able to select an engine at will taking in considerations what is compatible with its python version. It would be good for RADIS in terms of safety. It could also help to further increase the performances of RADIS if another package is better suited for our use.

### What has been done so far
_On the vaex side:_ 
- [ ] the vaex developer are working on it: https://github.com/vaexio/vaex/pull/2331

_On the RADIS side:_ 

- [ ] #656 allows to install RADIS with any Python version. However, if `vaex` is not compatible, the installation continues without vaex and `pytables` is the default engine.

### What are the alternatives

- [ ] [polar](https://pypi.org/project/polars/)               **===> preferred option <===**
Polars is a fast DataFrame library implemented in Rust and designed for high performance, with a strong focus on efficiency and speed. Thanks @ddelange for this suggestion in #656.

- [ ] [PySpark](https://spark.apache.org/docs/latest/api/python/)
PySpark is the Python API for Spark, a powerful open-source distributed computing system that can handle big data processing.

- [ ] [Dask](https://dask.org/)
Dask is a flexible parallel computing library for analytics that enables performance at scale. It integrates seamlessly with NumPy, pandas, and other Python libraries.

**Comment by erwanp** [comment]
Just a few comments.

- PySpark and Dask are distributed librairies; i.e. they scale by using more nodes on clusters. On a single laptop (which is what most RADIS users use) they won't help. 

 - I first heard of `polars` a couple years ago and I have a positive idea about what they're doing there; however, having researched it more today in June 2024 it seems they won't be a match for Vaex. This is because we're mostly using Vaex for read/write of our Dataframes; and our files are stored in HDF5 format. All benchmarks I could find was showing Polars to be ~3x faster than Pandas and ~2x faster than Pandas on reading CSV files... **But CSV-storage is very inefficient compared to HDF5**. Reading HDF5 can typically be 1000x faster than reading a csv. ([on a 20m dataset point](https://waterprogramming.wordpress.com/2023/06/22/intro-to-hdf5-h5py-and-comparison-to-csv-for-speed-compression/), reading HDF5 is 27x faster than reading CSV. On billion of lines datasets it's an even bigger difference.)   **So it's not so much about a choice of the library than a choice of the local data format we're using**.  ... and the problem is that Polars doesn't handle HDF5 format yet**  (https://github.com/pola-rs/polars/issues/3520)

Good summary of why HDF5 still seems to be the best choice : https://stackoverflow.com/questions/37928794/which-is-faster-for-load-pickle-or-hdf5-in-python .
- Parquet was fast ; but you must read the whole dataset : not good when you only need some spectral lines among a huge dataset. 


For whoever reads this threads : 

- I found an old Jupyter notebook comparing SQL vs HDF5 storage on HITEMP data. It could be adapted and re-run with some of the latest versions of the modern librairies (polars; vaex; pandas2; ) and file formats (parquet; arrow; hdf5) ; and see if there is not a better method to read line databases in 2024 : https://github.com/radis/radis-benchmark/blob/master/manual_benchmarks/test_database_loading.ipynb

**Comment by ddelange** [comment]
tl;dr: polars <3 arrow

> Parquet was fast ; but you must read the whole dataset

polars has lazy read variants [`scan_ipc`](https://docs.pola.rs/api/python/stable/reference/api/polars.scan_ipc.html) and [`scan_parquet`](https://docs.pola.rs/api/python/stable/reference/api/polars.scan_parquet.html) for this purpose.

nevertheless, both formats are column based, which is how arrow achieves such high compression ratios without sacrificing read speeds (especially feather v2 is [quite impressive](https://ursalabs.org/blog/2020-feather-v2/)). that also means however that you can't append a row to an existing file. if you dont mind that, i think performance-wise, polars+arrow will be a worthy contender even for row based operations, making it a worthwhile alternative standard file format to consider :)

**Comment by ddelange** [comment]
fwiw, in astronomy, images of spectroscopic sensors are stored in hdf5 as well. they're called 'cubes', where each pixel of the image contains a spectrum (different wavelengths of light). it's by far the fastest for slicing and dicing without loading the whole thing into memory. sounds like you're dealing with a very similar use-case here, so maybe there's no sense looking further than hdf5 indeed...

**Comment by erwanp** [comment]
> fwiw, in astronomy, images of spectroscopic sensors are stored in hdf5 as well. they're called 'cubes', where each pixel of the image contains a spectrum (different wavelengths of light). it's by far the fastest for slicing and dicing without loading the whole thing into memory. sounds like you're dealing with a very similar use-case here, so maybe there's no sense looking further than hdf5 indeed...

It is, and from your last comment I undertand than FeatherV2 is also a good option is we only need to slice but not append any new line  (which is what we do for most of our databases: we need partial access but we don't append/edit them anymore)

(@minouHub for information)

**Comment by khushi-070906** [comment]
@minouHub I have reviewed this issue, studied the current Vaex limitations with Python ≥3.11, and explored alternative engines. I am actively working on a solution and investigating Polars as a potential replacement within RADIS.

**Comment by UddhavB17** [comment]
Hi! I’d like to work on this for GSoC prep. 
Plan:
1. benchmark Vaex vs Polars/DuckDB/Dask on RADIS-like workloads, 
2. submit benchmark PR first, 
3. then integrate selected backend behind DATAFRAME_ENGINE without breaking existing users. 
Is this approach okay?

**Comment by Prtm2110** [comment]
@UddhavB17 this is part of the project therefore you can work on it once the project is selected. Feel free to submit proposal.

**Comment by ddelange** [comment]
fyi vaex 4.19 was released beginning of February with python 3.12 and numpy v2 support
